### PR TITLE
Run through Ember Build Pipeline

### PR DESCRIPTION
**Pull request by seanpdoyle**
* Share build process found in [emberjs/emberjs-build][1]
* Easiest approach was to copy paste from [emberjs/ember.js][2]
* In the long term, we should move common build tasks / setup to
[emberjs/emberjs-build][1]

[1]: https://github.com/emberjs/emberjs-build
[2]: https://github.com/emberjs/ember.js

### CODE REVIEW

**Comment by seanpdoyle** [comment]
This is currently a WIP

**Comment by seanpdoyle** [comment]
Current blocker: 


```
Uncaught Error: Could not find module ember-inflector/system/string
```

Ember Build can't accept additional ES6 modules (like those found here https://github.com/emberjs/emberjs-build/blob/6b68577851a92ab989754f93ad832c14b8a54cc9/lib/get-vendored-packages.js#L13-L17)

**Comment by bmac** [comment]
Hey @seanpdoyle whats the current status on this pr. Are you still blocked on importing ember-inflector?

**Comment by seanpdoyle** [comment]
@bmac yes, sort of. @twokul has recently refactored some of the emberjs-build project, so I have to catch up to the work he's done there. I might have some new issues stemming from his changes, but my biggest problem has been figuring out how to properly vendor / import / concat  `ember-data`, `activemodel_adapter`, and `ember-inflector`

**Comment by rwjblue** [comment]
@seanpdoyle - I'd be happy to do a hangout early next week to make a plan on this.

**Comment by seanpdoyle** [comment]
@rwjblue :+1: feel free to connect through email, hangouts, screenhero (all seandoyle at thoughtbot.com)

**Comment by seanpdoyle** [comment]
ping @rwjblue when do you have availability?

**Comment by seanpdoyle** [comment]
ping @rwjblue: do you have availability this week?

**Comment by seanpdoyle** [comment]
I'm running against https://github.com/seanpdoyle/emberjs-build/compare/wip-get-vendored-packages

and when I visit http://localhost:4200/tests/index.html and put a breakpoint on 

```javascript
 if (!registry[name]) {
  throw new Error("Could not find module " + name);
}
```

`ember-inflector` doesn't have a top level module. `ember-inflector/main` is the closest thing:

screen shot 2015-01-26 at 3 49 07 pm

cc: @twokul @rwjblue -- Any ideas?

**Comment by rwjblue** [comment]
@seanpdoyle - `ember-inflector/lib/main.js` should be moved to `ember-inflector.js` which would make it `ember-inflector` module. In other words `packages/<some-name>/lib/main.js` is `<some-name>` module.

**Comment by fivetanley** [comment]
i will gladly merge any commits to ember-inflector for it.

as for activemodel-adapter we have made plans to extract it into its own repository. We can start work on that and transfer it to the emberjs org later if that's more convenient.

**Comment by seanpdoyle** [comment]
@rwjblue do you mean in a build step, or do you mean in the actual source code at stefanpenner/ember-inflector?

**Comment by rwjblue** [comment]
@seanpdoyle - Build step.  https://github.com/emberjs/emberjs-build/blob/master/lib/htmlbars-package.js does roughly what you want (but you need to supply a `libPath` like in https://github.com/emberjs/emberjs-build/pull/63).

File: Brocfile.js [hunk 1]
```diff
@@ -1,174 +1,28 @@
...
+var emberBuild = new EmberBuild({
+  name: 'ember-data',
+  namespace: 'DS',
+  packages: packages,
+  skipTemplates: true,
+  skipRuntime: true,
+  vendoredPackages: {
+    'loader': vendoredPackage('loader'),
+    'ember-inflector': vendoredES6Package('ember-inflector', {
+      libPath: 'bower_components/ember-inflector/packages/ember-inflector/lib',
```
Comment (seanpdoyle) [inline Brocfile.js]: @rwjblue I thought this did what you're talking about in https://github.com/emberjs/data/pull/2687#issuecomment-71546143

Comment (rwjblue) [inline Brocfile.js]: Note that https://github.com/emberjs/emberjs-build/blob/master/lib/es6-vendored-package.js does not move `main.js` to `<package-name.js>` like the section I linked.

Comment (seanpdoyle) [inline Brocfile.js]: @rwjblue does this section https://github.com/emberjs/emberjs-build/blob/master/lib/htmlbars-package.js#L20-L28 do the renaming through the `includes` option for `Funnel`?

Comment (rwjblue) [inline Brocfile.js]: Crap, I'm making this as clear as mud.  This is what you need to move `main` up to the main package name: https://github.com/emberjs/emberjs-build/blob/master/lib/vendored-package.js#L28-L37

Comment (seanpdoyle) [inline Brocfile.js]: :ok_hand: that seems to have worked, but now 
screen shot 2015-01-26 at 5 20 45 pm

I can't find `qunit_configuration` and 0 tests are running.

I might just hold off until we have our hangout. I don't think  I fully understand all the moving parts

Comment (rwjblue) [inline Brocfile.js]: Basically this: screenshot

File: package.json [hunk 2]
```diff
@@ -18,42 +18,20 @@
...
+    "chalk": "~0.4.0",
     "defeatureify": "~0.1.4",
     "ejs": "^1.0.0",
-    "ember-cli": "^0.1.4",
+    "ember-cli": "~0.1.11",
     "ember-publisher": "0.0.6",
-    "es6-module-transpiler": "^0.9.5",
-    "es6-module-transpiler-amd-formatter": "^0.2.4",
-    "es6-module-transpiler-package-resolver": "^1.0.1",
+    "emberjs-build": "file:../emberjs-build",
```
Comment (seanpdoyle) [inline package.json]: @rwjblue this needs to be replaced, obviously, but this PR depends on https://github.com/emberjs/emberjs-build/pull/65

**Comment by seanpdoyle** [comment]
I'll try and tidy up the git history to tell a clearer story.

**Comment by seanpdoyle** [comment]
Heres the diff of `ember build --environment production`, with `master` on the left and this branch on the right

https://www.diffchecker.com/2nix5m8n

**Comment by twokul** [comment]
@seanpdoyle no source maps?

**Comment by seanpdoyle** [comment]
@rwjblue how does ember proper export source maps?

**Comment by rwjblue** [comment]
@seanpdoyle / @twokul - emberjs-build does not currently generate sourcemaps.  I am nearly finished with the migration to esperanto as the module transpiler, and we should be able to support sourcemaps for both projects once that effort is finished.

**Comment by seanpdoyle** [comment]
@rwjblue `ember build` is only outputting the `ember-data.debug.js`. It's skipping the `.prod` and `.min` versions of both `ember-data` and `ember-data-tests`.

I'm using [this version](https://github.com/emberjs/emberjs-build/blob/f801dc6e1e1ef286d49882e041f092a32252d9d6/lib/ember-build.js#L60-L66) of `ember-build`.

Any ideas?

File: bin/run-tests.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+testFunctions.push(function() {
+  return run('emberchannel=' + emberChannel);
+});
+testFunctions.push(function() {
+  return run('dist=min&emberchannel=' + emberChannel);
+});
+testFunctions.push(function() {
+  return run('dist=prod&emberchannel=' + emberChannel);
+});
```
Comment (seanpdoyle) [inline bin/run-tests.js]: @rwjblue I'm trying to setup `npm test` to run against `debug`, `prod` and `min` versions of the code.

 `ember build` is only outputting the `ember-data.debug.js`. It's skipping the `.prod` and `.min` versions of both `ember-data` and `ember-data-tests`.

I'm using [this version](https://github.com/emberjs/emberjs-build/blob/f801dc6e1e1ef286d49882e041f092a32252d9d6/lib/ember-build.js#L60-L66) of `ember-build`.

They exist when I run `ember build --environment production`, which is fine for CI where broccoli's environment is `production`, but locally only the first test run will work.

ember proper doesn't seem to be doing anything differently from the `ember-build` perspective.

Any ideas?

**Comment by seanpdoyle** [comment]
@rwjblue @twokul I think this is ready for a final look. Once I get the ok I'll squash into more meaningful commits

**Comment by rwjblue** [comment]
> ember build is only outputting the ember-data.debug.js. It's skipping the .prod and .min versions of both ember-data and ember-data-tests.

@seanpdoyle - It only generates prod and min output when running against the production environment (this ensures that dev builds are MUCH faster). So `ember build --environment=production` should do the trick here.

**Comment by seanpdoyle** [comment]
Looks like this is failing against canary because of too many deprecation warnings logged.

https://travis-ci.org/emberjs/data/jobs/48790205

Should this PR address that?

**Comment by wecc** [comment]
Does this PR also fix #2753?

**Comment by wecc** [comment]
This should close #2683

**Comment by seanpdoyle** [comment]
@wecc that is correct.

I'm waiting on a final approval before I squash to fewer commits.

@rwjblue @twokul how does this look?

**Comment by bmac** [comment]
@seanpdoyle the tests failing in canary should be fixed now if you rebase against master.

**Comment by seanpdoyle** [comment]
@bmac now the deprecation warnings have spread to beta (https://travis-ci.org/emberjs/data/jobs/49986468).

**Comment by seanpdoyle** [comment]
This is now failing / erroring against `stable` due to [this line](https://github.com/emberjs/ember.js/blob/729e2cd19a92958d24d35dbabcf3faf55fd9d70e/packages/loader/lib/main.js#L42)

Also, thanks to ~5000 deprecation warnings, it errors out / times out in CI against `beta` and `canary`.

The `container-deprecations` branch (found [here](https://github.com/seanpdoyle/data/compare/container-deprecations)) addresses the deprecations, but breaks the build against `stable`.

@rwjblue do you have any suggestions?

**Comment by bmac** [comment]
@seanpdoyle is `env.registry` undefined with stable? I wonder if we could feature detect stable do something like `env.registry = container` in https://github.com/emberjs/data/blob/master/tests/ember_configuration.js for to get the stable branch passing.

**Comment by seanpdoyle** [comment]
@bmac good idea. Changes in https://github.com/emberjs/data/pull/2773

**Comment by seanpdoyle** [comment]
@rwjblue any ideas / suggestions on why `stable` can no longer make it past this call in the [loader](https://github.com/emberjs/ember.js/blob/729e2cd19a92958d24d35dbabcf3faf55fd9d70e/packages/loader/lib/main.js#L42)?

```js
var value = length === 0 ? callback.call(this) : callback.apply(this, reified);
```

It's failing for modules that aren't exporting callbacks:

* [Ember.Date](https://github.com/emberjs/data/blob/f60e42905df184d1e644a6cda8443a8156bc8990/packages/ember-data/lib/ext/date.js) since it isn't exporting anything
* modules containing in `.jscs-test`, since I'm assuming they're not exporting anything
* modules containing in `jshint`, since I'm assuming they're not exporting anything.

**Comment by bmac** [comment]
ping @seanpdoyle what is the current status of this pr? Is there anything you are blocked on?

**Comment by seanpdoyle** [comment]
@bmac https://github.com/emberjs/data/pull/2687#issuecomment-73965469 for one, and https://github.com/ember-cli/ember-cli/commit/5d10affbfeea696b1db9f98c58c816bae9ebac89#commitcomment-9915930

**Comment by bmac** [comment]
@seanpdoyle I did some poking around in this pr today. It looks only the index.html file is being coppied form the tests/ directory. I opened a pr to try and get Ember Data's .js files included. https://github.com/emberjs/emberjs-build


For Ember.Date we could try and export `Ember.Date` to make the build happy.


For the `.jscs-test` and `jshint` tests it looks like the problem is the `Ember.__loader.define` method in the release ember doesn't support AMD modules defined without the dependencies argument `define('.jscs-test', [], function() {})` vs `define('.jscs-test', function() {})`. The jscs and jshint modules are generating tests that match the 2nd style. 

It looks like @rwjblue has fixed this bug in Ember 1.11. https://github.com/rwjblue/ember.js/commit/3bb454762ce7d62f03bb6f778f9a3831829b72fa

**Comment by seanpdoyle** [comment]
@bmac thanks for digging into this.

I've tried again with your emberjs-build branch. That seems to have fixed the loading issue.

the `Ember.__loader` issues aren't a problem with `beta` or `canary`. @rwjblue is there anyway to backport https://github.com/rwjblue/ember.js/commit/3bb454762ce7d62f03bb6f778f9a3831829b72fa to the `release` build?

New issues:

* JSHint failures galore
* screen shot 2015-03-01 at 9 58 23 pm tons of deprecation warnings stemming from https://github.com/emberjs/data/blob/b8aff0910775f864d6f918ecda1333491a3c001f/packages/ember-data/lib/system/store.js#L1437 calling a method on `this.container` instead of a `Registry` instance
* travis can't build the project for the tests: https://travis-ci.org/emberjs/data/jobs/52689184

**Comment by seanpdoyle** [comment]
@rwjblue @bmac this is green on `beta` and `canary` (https://travis-ci.org/emberjs/data/builds/53127074)

What, if any is the best approach for backporting https://github.com/rwjblue/ember.js/commit/3bb454762ce7d62f03bb6f778f9a3831829b72fa?

**Comment by seanpdoyle** [comment]
@igorT @fivetanley @rwjblue how should I go about fixing this against release?

Is it better / easier for me to just wait until `beta` becomes `release`?

**Comment by rwjblue** [comment]
It is possible to monkey patch Ember.__loader.define (save the original, override with a copy that adds the missing second param as an empty array, then call the original function with the correct args), or you can wait a couple weeks until release has the fix.

Totally up to you...

**Comment by seanpdoyle** [comment]
@rwjblue where in the lifecycle could I apply that monkey patch?

I'm making that call [here](https://github.com/emberjs/data/blob/f480016859746badae064546186eeab631e46745/packages/ember-data/lib/main.js#L6) and it's not early enough, as that file is wrapped in a `define` callback.

**Comment by seanpdoyle** [comment]
@rwjblue also, I can't believe I've never asked this, but is the monkey patching a test-only fix?

**Comment by seanpdoyle** [comment]
@rwjblue how would I go about changing `emberjs-build` to include `bower_components` so that this line

https://github.com/emberjs/data/blob/6990154edccbccd2e695248c1f0a1eaea3305c89/tests/index.html#L7

loads `qunit.js` et al.?

I tried adding `/bower_components/` to https://github.com/emberjs/emberjs-build/blob/e13695255dcc33ecabd509bd5d719a54e961335d/lib/config/test-config-tree.js#L11 but it wasn't included

**Comment by rwjblue** [comment]
@seanpdoyle - It is available at `/qunit/qunit.js` by default, you should be able to update that script tag to something like:

```
<script src="../qunit/qunit.js"></script>
```

**Comment by seanpdoyle** [comment]
this is green depending on https://github.com/emberjs/emberjs-build/pull/89 being merged

**Comment by seanpdoyle** [comment]
@rwjblue this is green depending on https://github.com/emberjs/emberjs-build/pull/89 being merged

**Comment by rwjblue** [comment]
@seanpdoyle - I do not believe that an update to emberjs-build is needed (at least, not to get qunit to load).  Did you see my comment above?

**Comment by seanpdoyle** [comment]
@rwjblue I did, but ember data needs (wants?) the `ember` bower component.

We could probably get by without it, but that means we'd pull from `builds.emberjs.com` instead of the local file system

**Comment by rwjblue** [comment]
> that means we'd pull from builds.emberjs.com instead of the local file system

Which unfortunately is what is happening today.


I agree that changes are needed, sorry for the quick reply without digging into the reasoning more.  I commented on the emberjs-build pull request with some suggestions.

**Comment by seanpdoyle** [comment]
@rwjblue this is green for `release`, but for `beta` and `canary`, deprecation warnings are flooding travis and failing an otherwise green build.

https://travis-ci.org/emberjs/data/jobs/56345865#L229

> DEPRECATION: Using the same function as getter and setter is deprecated. See http://emberjs.com/deprecations/v1.x/#toc_deprecate-using-the-same-function-as-getter-and-setter-in-computed-properties for more details.

This change is introduced in `1.12`, so I can't introduce the fix without breaking `release`.

Is there a way to silence deprecations?

**Comment by seanpdoyle** [comment]
@rwjblue @fivetanley @bmac ping

**Comment by rwjblue** [comment]
@seanpdoyle the deprecations are fixed on master in https://github.com/emberjs/data/pull/2961 but that is not what was causing it to fail. The actual issue which was causing the failure was fixed in https://github.com/emberjs/data/pull/2959.

Can you rebase to get these updates?

**Comment by seanpdoyle** [comment]
@rwjblue what a time to be alive

**Comment by seanpdoyle** [comment]
Looks like the windows build is failing because it's using node `0.12.x` (https://ci.appveyor.com/project/embercli/data/build/724#L325) instead of `0.10.x`

**Comment by rwjblue** [comment]
Confirm that issue is being tracked in https://github.com/emberjs/emberjs-build/issues/87, but I am unsure if @stefanpenner has had time to work on it more.

**Comment by seanpdoyle** [comment]
note to self:

squashed and force pushed over

```diff
(function() {
+ window.EmberDev = window.EmberDev || {};
+ EmberDev.runningProdBuild = !!QUnit.urlParams.prod;
+ // hack qunit to not suck for Ember objects
+ var originalTypeof = QUnit.jsDump.typeOf;
+ QUnit.jsDump.typeOf = function(obj) {
+ if (Ember && Ember.Object && Ember.Object.detectInstance(obj)) {
+ return "emberObject";
+ }
+ return originalTypeof.call(this, obj);
+ };
+ // raises is deprecated, but we likely want to keep it around for our es3
+ // test runs.
+ QUnit.constructor.prototype.raises = QUnit['throws'];
+ window.raises = QUnit['throws'];
+ QUnit.jsDump.parsers.emberObject = function(obj) {
+ return obj.toString();
+ };
+ var EmberDevTestHelperAssert = window.Ember.__loader.require('ember-dev/test-helper/index')['default'];
+ var setupQUnit = window.Ember.__loader.require('ember-dev/test-helper/setup-qunit')['default'];
+ var testHelpers = new EmberDevTestHelperAssert(window.Ember, EmberDev.runningProdBuild);
+ setupQUnit(testHelpers, QUnit);
+ window.module = QUnit.module;

+ // Tests should time out after 5 seconds
+ QUnit.config.testTimeout = 5000;
+ // Hide passed tests by default
+ QUnit.config.hidepassed = true;
```

**Comment by seanpdoyle** [comment]
@stefanpenner still getting the same node `0.12.x` Windows failure.

Has the landscape changed at all in the last few months since I've worked on this?

**Comment by rwjblue** [comment]
@seanpdoyle - Not that I am aware of.

**Comment by seanpdoyle** [comment]
@rwjblue I would love to get this merged during this pre-June 12th push.

**Comment by seanpdoyle** [comment]
@rwjblue @stefanpenner new and exciting failure:

```
› node -v
v0.10.38
› ember test
Future versions of Ember CLI will not support v0.10.38. Please update to Node 0.12 or io.js.
version: 0.2.7
Built project successfully. Stored in "/Users/Sean/src/ember-data/tmp/class-tests_dist-PTuvffjr.tmp".
dyld: lazy symbol binding failed: Symbol not found: _node_module_register
  Referenced from: /Users/Sean/src/ember-data/node_modules/ember-cli/node_modules/testem/node_modules/socket.io/node_modules/engine.io/node_modules/ws/build/Release/bufferutil.node
  Expected in: dynamic lookup

dyld: Symbol not found: _node_module_register
  Referenced from: /Users/Sean/src/ember-data/node_modules/ember-cli/node_modules/testem/node_modules/socket.io/node_modules/engine.io/node_modules/ws/build/Release/bufferutil.node
  Expected in: dynamic lookup

zsh: trace trap  ember test
```

**Comment by wecc** [comment]
I'm gonna go ahead and close this since ED now has been addonified.

@seanpdoyle Really appreciate your work on this even though it didn't get merged :+1:

**Comment by seanpdoyle** [comment]
:clap: I'm glad that's the direction this library has gone.

**Pull request closed without merging**